# Major League Baseball Home Run Exit Velocities

One of the most recent debates in Major League Baseball is focused on why more home runs were hit in the 2017 season than any other season in the leagues history. There were 6,105 home runs hit in the 2017 season, which is more home runs than in the peak of the steroid era and everyone wants to know why. One specific area of interest for investigation is related to the most important item to the game, the baseball. During the 2017 season there were numerous complaints from major league pitchers that the ball felt different and the result was a record breaking year for home runs.

Thanks to Baseball Savant and their StatCast technology we can begin to understand what is happening by analyzing the data tracked on every pitch that was hit for a home run during these seasons. After a high level review of the data and some outside research I found a specific stat known as the Home Run Exit Velocity. Home Run Exit Velocity as defined by Baseball Savant is the speed at which the baseball leaves the bat after being hit and it would make sense that an increase in speed would lead to further distance and thus yield a home run. 

So, using machine learning modeling techniques I will build a production level regression model to draw a conclusion on what features involved in a home run are most influential to the home run exit velocity. The data that will be used to train this model will come from three different aspects:
1. StatCast pitch data for all home run hits
2. Sample of baseballs used in the 2015, 2016 and 2017 seasons
3. Baseball player personal stats (height, weight, age)

From this model I will draw a conclusion about what are the most influential features contributing to a batters home run exit velocity.

In [1]:
import pandas as pd
import numpy as np
from scipy import stats 

from bs4 import BeautifulSoup
import requests
import json
import time

## Data Cleaning

### Import the Data

The following notebook will load in two datasets for review and cleaning:
1. Each home run hit during the 2015, 2016 and 2017 MLB seasons
2. Stats from a sample of the actual baseballs taken from the 2015, 2016 and 2017 MLB seasons
3. I will also perform a webscrape of the website Baseball Savant for each players personal physical stats that hit a home run in the 2015, 2016 and 2017 seasons.
    - Height
    - Weight
    - Age
    
URL : https://baseballsavant.mlb.com/

#### Home Runs

To start I will review and clean the largest of my three datasets which is the statistical data on each home run hit during the 2015, 2016 and 2017 seasons.

In [2]:
hr_df = pd.read_csv('../data/home_runs_15_16_17.csv')
baseballs_df = pd.read_excel('../data/baseballs.xlsx')

In [3]:
hr_df.columns = [x.lower().replace(' ', '_') for x in hr_df.columns]

In [4]:
hr_df.head()

pitch_type game_date  release_speed  release_pos_x  release_pos_z  \
0         CH   6/10/17           84.5        -3.4020         6.4696   
1         FF   4/28/17           97.1        -2.8091         5.9279   
2         CU   6/23/15           86.7        -1.5647         5.3406   
3         SL   9/28/17           89.5         2.0682         6.1177   
4         SL   6/11/17           84.7        -1.9795         5.6860   

         player_name  batter  pitcher    events          description  \
0        Aaron Judge  592450   501957  home_run  hit_into_play_score   
1        Aaron Judge  592450   592332  home_run  hit_into_play_score   
2  Giancarlo Stanton  519317   593372  home_run  hit_into_play_score   
3  Giancarlo Stanton  519317   571521  home_run  hit_into_play_score   
4        Aaron Judge  592450   548337  home_run  hit_into_play_score   

           ...           home_score  away_score  bat_score  fld_score  \
0          ...                    0           0          0          0   
1          ...                    2           9          2          9   
2          ...                    0           0          0          0   
3          ...                    5           1          5          1   
4          ...                    7           3          7          3   

   post_away_score post_home_score post_bat_score post_fld_score  \
0                0               0              0              0   
1                9               2              2              9   
2                0               0              0              0   
3                1               5              5              1   
4                3               7              7              3   

  if_fielding_alignment of_fielding_alignment  
0         Infield shift              Standard  
1              Standard              Standard  
2              Standard              Standard  
3         Infield shift              Standard  
4         Infield shift              Standard  

[5 rows x 89 columns]

In [5]:
hr_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16626 entries, 0 to 16625
Data columns (total 89 columns):
pitch_type                         16602 non-null object
game_date                          16626 non-null object
release_speed                      16611 non-null float64
release_pos_x                      16502 non-null float64
release_pos_z                      16502 non-null float64
player_name                        16626 non-null object
batter                             16626 non-null int64
pitcher                            16626 non-null int64
events                             16626 non-null object
description                        16626 non-null object
spin_dir                           0 non-null float64
spin_rate_deprecated               0 non-null float64
break_angle_deprecated             0 non-null float64
break_length_deprecated            0 non-null float64
zone                               16611 non-null float64
des                                16626 non-n

#### Remove unnecessary features (noise)

After a review of all the available features within this dataset the following will be dropped:

**NOTE:** An in depth review of each of the columns can be found at basball savant: https://baseballsavant.mlb.com/

- Dropped because of lack of data:
    - `spin_dir`
    - `spin_rate_deprecated`
    - `break_angle_deprecated`
    - `break_length_deprecated` 
    - `hit_location`
    - `tfs_deprecated`
    - `tfs_zulu_deprecated`
    - `iso_value`
    - `launch_speed_angle`: the values of 1-6 are unbalanced and I cannot find the exact meanings. I already have seperate features for both launch angle and launch speed.
- Dropped because of no relation to the actual home run stats (launch speed, launch angle, etc.)
    - `on_3b`
    - `on_2b`
    - `on_1b`
    - `umpire`
    - `events`
    - `description`
    - `des`
    - `game_type`
    - `stand`
    - `type`
    - `balls`
    - `strikes`
    - `outs_when_up`
    - `inning`
    - `inning_topbot`
    - `fielder_2`
    - `sv_id`
    - `pitcher.1`
    - `fielder_2.1`
    - `fielder_3`
    - `fielder_4`
    - `fielder_5`
    - `fielder_6`
    - `fielder_7` 
    - `fielder_8 `
    - `fielder_9`
    - `estimated_ba_using_speedangle`
    - `estimated_woba_using_speedangle`
    - `woba_value`
    - `woba_denom` 
    - `babip_value`
    - `at_bat_number`
    - `pitch_number `
    - `home_score` 
    - `away_score` 
    - `bat_score`
    - `fld_score`
    - `post_away_score`
    - `post_home_score`
    - `post_bat_score`
    - `post_fld_score`
    - `if_fielding_alignment `
    - `of_fielding_alignment`
    - `game_pk`
    - `pitch_name` (duplicate to pitch type)
    - `home_team`
    - `away_team`
    - `hc_x : coordinate of where the ball is fielded`
    - `hc_y : coordinate of where the ball is fielded`

In [6]:
col = [ 
    'umpire',
    'spin_dir',
    'spin_rate_deprecated',
    'break_angle_deprecated',
    'break_length_deprecated',
    'events',
    'description', 
    'des',
    'game_type', 
    'stand',
    'type', 
    'hit_location', 
    'balls', 
    'strikes', 
    'on_3b', 
    'on_2b', 
    'on_1b', 
    'outs_when_up',
    'inning',
    'inning_topbot',
    'tfs_deprecated', 
    'tfs_zulu_deprecated', 
    'fielder_2', 
    'sv_id', 
    'pitcher.1', 
    'fielder_2.1', 
    'fielder_3', 
    'fielder_4', 
    'fielder_5', 
    'fielder_6', 
    'fielder_7', 
    'fielder_8', 
    'fielder_9', 
    'estimated_ba_using_speedangle', 
    'estimated_woba_using_speedangle', 
    'woba_value', 
    'woba_denom', 
    'babip_value', 
    'at_bat_number', 
    'pitch_number', 
    'home_score', 
    'away_score', 
    'bat_score', 
    'fld_score', 
    'post_away_score', 
    'post_home_score', 
    'post_bat_score', 
    'post_fld_score', 
    'if_fielding_alignment', 
    'of_fielding_alignment',
    'iso_value',
    'game_pk',
    'pitch_name',
    'away_team',
    'home_team',
    'hc_x',
    'hc_y',
    'launch_speed_angle'
]

In [7]:
hr_df.drop(col, axis=1, inplace=True)

#### Removing nulls

One of the features that is a very important aspect to a specific players home run stats is the pitch in which they hit a home run. So, I will be dropping the rows in which have nulls documented within the pitch type that was thrown.

In [8]:
col=['pitch_type']
hr_df.dropna(subset=col, inplace=True)

#### Updating datatypes

Ensuring specific data types are captured in the correct format

In [9]:
hr_df['game_date'] = pd.to_datetime(hr_df['game_date'])

In [10]:
hr_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16602 entries, 0 to 16625
Data columns (total 31 columns):
pitch_type           16602 non-null object
game_date            16602 non-null datetime64[ns]
release_speed        16600 non-null float64
release_pos_x        16491 non-null float64
release_pos_z        16491 non-null float64
player_name          16602 non-null object
batter               16602 non-null int64
pitcher              16602 non-null int64
zone                 16600 non-null float64
p_throws             16602 non-null object
bb_type              16602 non-null object
game_year            16602 non-null int64
pfx_x                16600 non-null float64
pfx_z                16600 non-null float64
plate_x              16600 non-null float64
plate_z              16600 non-null float64
vx0                  16600 non-null float64
vy0                  16600 non-null float64
vz0                  16600 non-null float64
ax                   16600 non-null float64
ay            

In [11]:
hr_df['pitch_type'].isnull().sum()

0

#### Filling in Nulls (Pitch Stats)

A lot of the pitching data related to the actual pitch that was thrown is missing. Using a groupby and a function I can fill these values in for the specific pitcher that threw the pitch. 

Steps to fill in these values:
1. Create a dataframe grouped by pitcher and pitch type to get an average for each pitch statistic
2. Create a function that will look up this picher by their ID and then look up their average stat from the newly created dataframe to fill in whatever missing values are missing for the specific pitch record
3. Apply the newly created function to the pitching features where nulls were identified

In [12]:
hr_df.isnull().sum()[['release_speed', 
                      'release_pos_x', 
                      'release_pos_y',
                      'release_pos_z',
                      'release_spin_rate',
                      'release_extension',
                      'pfx_x',
                      'pfx_z',
                      'plate_x',
                      'vx0',
                      'vy0',
                      'vz0',
                      'ax',
                      'ay',
                      'az',
                      'effective_speed']]

release_speed           2
release_pos_x         111
release_pos_y         111
release_pos_z         111
release_spin_rate    1011
release_extension     305
pfx_x                   2
pfx_z                   2
plate_x                 2
vx0                     2
vy0                     2
vz0                     2
ax                      2
ay                      2
az                      2
effective_speed       330
dtype: int64

In [14]:
pitch_avg = hr_df.groupby(['pitcher', 'pitch_type'])[['release_speed', 
                                                      'release_pos_x', 
                                                      'release_pos_y',
                                                      'release_pos_z',
                                                      'release_spin_rate',
                                                      'release_extension',
                                                      'pfx_x',
                                                      'pfx_z',
                                                      'plate_x',
                                                      'vx0',
                                                      'vy0',
                                                      'vz0',
                                                      'ax',
                                                      'ay',
                                                      'az',
                                                      'effective_speed']].mean()

In [15]:
pitch_avg.reset_index(inplace=True)

In [16]:
pitch_avg.head()

pitcher pitch_type  release_speed  release_pos_x  release_pos_y  \
0   112526         CH      81.540000      -1.391520      55.250980   
1   112526         FF      91.388235      -0.988865      55.251547   
2   112526         FT      87.585106      -1.425957      55.216726   
3   112526         SL      81.700000      -1.545200      55.437425   
4   115629         CH      84.300000      -1.830700      54.319700   

   release_pos_z  release_spin_rate  release_extension     pfx_x     pfx_z  \
0       5.754080        1637.600000           5.249200 -1.142933  0.835700   
1       5.992818        2276.562500           5.257500 -0.717337  1.424175   
2       5.713583        2109.446809           5.283426 -1.339216  0.963503   
3       5.835250        2309.000000           5.061250  0.077197  0.502796   
4       6.365500        1669.000000           6.148000 -1.051283  1.365300   

    plate_x       vx0         vy0       vz0         ax         ay         az  \
0 -0.223840  4.451960 -118.456220 -2.985940 -10.667300  22.154580 -25.151200   
1 -0.022853  3.560453 -132.811671 -4.633582  -8.148265  29.473624 -16.233365   
2 -0.173119  5.304594 -127.276523 -3.549502 -14.556315  26.805947 -22.843936   
3  0.172225  3.291075 -118.803012 -1.377837   1.219612  21.471113 -28.542300   
4 -0.379000  4.877000 -122.597000 -4.700000 -10.074000  24.338000 -19.731000   

   effective_speed  
0        79.894600  
1        89.426625  
2        85.687319  
3        79.894375  
4        83.946000

In [17]:
def pitcher_avg(pitch, col):
    return float(pitch_avg[(pitch_avg['pitcher']==pitch.pitcher) & 
                           (pitch_avg['pitch_type']==pitch.pitch_type)][col])

In [18]:
def pitch_stats_apply(columns):
    for col in columns:
        hr_df[col] = hr_df.apply(lambda x: pitcher_avg(x, col) if pd.isnull(x[col]) else x[col],1)
    return hr_df

In [19]:
columns = pitch_avg
hr_df = pitch_stats_apply(columns)

In [21]:
hr_df['zone'].iloc[9248] = 4

In [22]:
hr_df['zone'].iloc[9338] = 4

In [23]:
hr_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16602 entries, 0 to 16625
Data columns (total 31 columns):
pitch_type           16602 non-null object
game_date            16602 non-null datetime64[ns]
release_speed        16602 non-null float64
release_pos_x        16596 non-null float64
release_pos_z        16596 non-null float64
player_name          16602 non-null object
batter               16602 non-null int64
pitcher              16602 non-null int64
zone                 16602 non-null float64
p_throws             16602 non-null object
bb_type              16602 non-null object
game_year            16602 non-null int64
pfx_x                16602 non-null float64
pfx_z                16602 non-null float64
plate_x              16602 non-null float64
plate_z              16600 non-null float64
vx0                  16602 non-null float64
vy0                  16602 non-null float64
vz0                  16602 non-null float64
ax                   16602 non-null float64
ay            

#### Filling in Nulls (Home Run Hit Stats)

With as many nulls as possible now filled in for the missing pitch stats I will now perform the same steps for the missing home run hit stats. 

Steps to fill in these values:
1. Create a dataframe grouped by batter, the pitch type and the batted ball type to get an average for each for the batting statistics that contain nulls
2. Create a function that will look up the batter by their ID and then look up their average stat from the newly created dataframe to fill in missing values
3. Apply the newly created function to the batter features where nulls were identified

In [24]:
batter_stats = hr_df.groupby(['pitch_type', 'batter', 'bb_type']).mean()[['hit_distance_sc']]

In [25]:
batter_stats.reset_index(inplace=True)

In [26]:
batter_stats.head()

pitch_type  batter     bb_type  hit_distance_sc
0         CH  116338    fly_ball       391.666667
1         CH  120074    fly_ball       379.000000
2         CH  120074  line_drive       413.666667
3         CH  121347    fly_ball       385.500000
4         CH  121347  line_drive       412.000000

In [27]:
def batter_avg(pitch, col):
    return float(batter_stats[(batter_stats['batter']==pitch.batter) &
                              (batter_stats['pitch_type']==pitch.pitch_type) &
                              (batter_stats['bb_type']==pitch.bb_type)][col])

In [28]:
def batter_stats_apply(columns):
    for col in columns:
        hr_df[col] = hr_df.apply(lambda x: batter_avg(x, col) 
                                               if pd.isnull(x[col])
                                               else x[col],1)

In [29]:
columns = batter_stats
batter_stats_apply(columns)

In [30]:
hr_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16602 entries, 0 to 16625
Data columns (total 31 columns):
pitch_type           16602 non-null object
game_date            16602 non-null datetime64[ns]
release_speed        16602 non-null float64
release_pos_x        16596 non-null float64
release_pos_z        16596 non-null float64
player_name          16602 non-null object
batter               16602 non-null int64
pitcher              16602 non-null int64
zone                 16602 non-null float64
p_throws             16602 non-null object
bb_type              16602 non-null object
game_year            16602 non-null int64
pfx_x                16602 non-null float64
pfx_z                16602 non-null float64
plate_x              16602 non-null float64
plate_z              16600 non-null float64
vx0                  16602 non-null float64
vy0                  16602 non-null float64
vz0                  16602 non-null float64
ax                   16602 non-null float64
ay            

#### Final missing value fill

I have now filled in as many nulls as possible within the dataset and will now drop all remaining null values

In [31]:
hr_df.dropna(inplace=True)

In [32]:
len(hr_df)

16388

### Save the Data

In [33]:
hr_df.to_csv('../data/final_clean_hr.csv')

### Baseball Data

Next, I will review and clean the sampled baseballs from the 2015, 2016 and 2017 MLB seasons

In [34]:
baseballs_df.columns = [i.lower().replace(' ', '_') for i in baseballs_df.columns]

In [35]:
baseballs_df.head()

ball_code       year   sn  weight_(oz)  circumference_(in)  avg_seam_height  \
0  MSCC0051 2014-05-15  731        5.135                9.13          0.03587   
1  MSCC0032 2014-07-15  228        5.149                9.09          0.04403   
2  MSCC0030 2015-04-15  196        5.143                9.06          0.03726   
3  MSCC0045 2015-04-15  351        5.109                9.09          0.04574   
4  MSCC0048 2015-04-15  499        5.192                9.19          0.04862   

   std_of_seam_height  avg_ccor  avg_ds  #_of_good_shots_before_damage  
0             0.01049     0.491   12133                            NaN  
1             0.01781     0.489   12468                            NaN  
2             0.00640     0.489   12518                            NaN  
3             0.01216     0.496   13442                            NaN  
4             0.00828     0.474   12394                            5.0

#### Dropping unnecessary features (noise)
- Dropping because of their irrelevance to home run statistics
    - ball code
    - sn
- Dropping because of their lack of data
    - #_of_good_shots_before_damage

In [36]:
col = [
    'ball_code',
    '#_of_good_shots_before_damage',
    'sn'
]

baseballs_df.drop(col, axis=1, inplace=True)

In [37]:
baseballs_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36 entries, 0 to 35
Data columns (total 7 columns):
year                  36 non-null datetime64[ns]
weight_(oz)           36 non-null float64
circumference_(in)    36 non-null float64
avg_seam_height       36 non-null float64
std_of_seam_height    36 non-null float64
avg_ccor              36 non-null float64
avg_ds                36 non-null int64
dtypes: datetime64[ns](1), float64(5), int64(1)
memory usage: 2.0 KB


#### Updating datatypes 

Updating fetures to their correct data types and updating the index for this dataframe to datetime for efficiency

In [38]:
baseballs_df['year'] = pd.to_datetime(baseballs_df['year'])

In [39]:
baseballs_df.set_index('year', inplace=True)

In [40]:
baseballs_df.head()

weight_(oz)  circumference_(in)  avg_seam_height  \
year                                                           
2014-05-15        5.135                9.13          0.03587   
2014-07-15        5.149                9.09          0.04403   
2015-04-15        5.143                9.06          0.03726   
2015-04-15        5.109                9.09          0.04574   
2015-04-15        5.192                9.19          0.04862   

            std_of_seam_height  avg_ccor  avg_ds  
year                                              
2014-05-15             0.01049     0.491   12133  
2014-07-15             0.01781     0.489   12468  
2015-04-15             0.00640     0.489   12518  
2015-04-15             0.01216     0.496   13442  
2015-04-15             0.00828     0.474   12394

#### Creating a final average of the baseballs to use for each MLB season

- 2015
    - Get the average of each statistic from the 2015 season and create as a dataframe
- 2016
    - Get the average of each statistic from the 2016 season and create as a dataframe
- 2017
    - It was extremely difficult to find the data for the baseballs from the 2017 season. Using a couple of articles online I was able to construct some averages for 2017
        - Sources: 
            - https://www.theringer.com/2017/6/14/16044264/2017-mlb-home-run-spike-juiced-ball-testing-reveal-155cd21108bc 
            - https://fivethirtyeight.com/features/juiced-baseballs/

In [41]:
bb_15 = baseballs_df['2015'][['weight_(oz)',
                      'circumference_(in)',
                      'avg_seam_height',
                      'std_of_seam_height',
                      'avg_ccor',
                      'avg_ds']].mean().to_frame().T
bb_15['year'] = 2015
bb_15

weight_(oz)  circumference_(in)  avg_seam_height  std_of_seam_height  \
0     5.120125            9.107917         0.040502            0.011194   

   avg_ccor     avg_ds  year  
0  0.491333  12740.875  2015

In [42]:
bb_16 = baseballs_df['2016'][['weight_(oz)',
                      'circumference_(in)',
                      'avg_seam_height',
                      'std_of_seam_height',
                      'avg_ccor',
                      'avg_ds']].mean().to_frame().T
bb_16['year'] = 2016
bb_16

weight_(oz)  circumference_(in)  avg_seam_height  std_of_seam_height  \
0       5.1225               9.079         0.038432            0.010468   

   avg_ccor   avg_ds  year  
0    0.4941  12926.4  2016

In [43]:
mlb_baseballs = bb_15.append(bb_16).reset_index()
mlb_baseballs.drop('index', axis=1, inplace=True)
mlb_baseballs.drop('std_of_seam_height', axis=1, inplace=True)
mlb_baseballs

weight_(oz)  circumference_(in)  avg_seam_height  avg_ccor     avg_ds  year
0     5.120125            9.107917         0.040502  0.491333  12740.875  2015
1     5.122500            9.079000         0.038432  0.494100  12926.400  2016

#### The 2017 baseballs

The largest point of emphasis that was examined within the baseballs was the weight and ccor (cylindrical coefficient of restitution). Because these are the main data points I will focus on updating these values for the 2017 baseballs.

"The newer cores weigh about a half a gram less than the older ones, which might be enough to cause baseballs hit on a typical home run trajectory to fly about 6 inches farther." -FiveThirtyEight

- Weight
        - 5.12250 - 0.017637 (.5 grams converted to ouces) = 5.104863
- Avg ccor (Ccylindrical coefficient of restitution) "The bounciness of the baseball aka "The Pill""
        - 0.494100 - 0.017637 (.5 grams converted to ouces) = .476463

In [44]:
bb_17 = pd.DataFrame(
    [[.476463,.038,9.08,5.104863,12926,2017]], columns=['avg_ccor', 
                                                   'avg_seam_height',
                                                   'circumference_(in)',
                                                   'weight_(oz)',
                                                   'avg_ds',
                                                   'year']
)
bb_17

avg_ccor  avg_seam_height  circumference_(in)  weight_(oz)  avg_ds  year
0  0.476463            0.038                9.08     5.104863   12926  2017

In [45]:
mlb_baseballs = mlb_baseballs.append(bb_17, sort=False).reset_index().drop('index', axis=1)

In [46]:
mlb_baseballs

weight_(oz)  circumference_(in)  avg_seam_height  avg_ccor     avg_ds  year
0     5.120125            9.107917         0.040502  0.491333  12740.875  2015
1     5.122500            9.079000         0.038432  0.494100  12926.400  2016
2     5.104863            9.080000         0.038000  0.476463  12926.000  2017

### Save the Data

In [47]:
mlb_baseballs.to_csv('../data/final_mlb_baseballs.csv')

### Player Personal Stats

Finally, I want to include each batters personal stats to draw a conclusion about whether or not a players size and weight are contributing to their home runs exit velocities.

#### Webscapre Baseball Savant

In order to scrape a website an appropriate status code must be received. Using a player at random I will send a request to the website to ensure a successful connection.
- I want to receive a status code of 200

In [67]:
url = "https://baseballsavant.mlb.com/savant-player/j-d-martinez-502110?stats=career-r-hitting-mlb"
res =  requests.get(url)

In [68]:
if res.status_code == 200:
    print(f'Successfull status code received : {res.status_code}')

Successfull status code received : 200


#### The webscrape

Lets ensure we know how many unique players we'll need to scrape. 

There are 751 players that have hit homes runs since 2015. I will scrape each of these players personal stats:
- Height
- Weight
- Age

In [50]:
batter_id = hr_df['batter'].unique()

In [51]:
len(batter_id)

751

#### Running the scrape

Using Pythons Beautiful Soup Library I will run a loop through each players personal page on the Baseball Savant site and obtain their speicific personal stats.

In [52]:
def get_batters_info(batter_id):
    attrs = {
        'batter_id': [str(x) for x in batter_id],
        'height':[],
        'weight':[],
        'age':[],
        'position':[],
        'bats':[],
        'throws':[]
    }

    attr_names = [
        'height',
        'weight',
        'age',
        'position',
        'bats',
        'throws'
    ]

    for idx, b in enumerate(batter_id):
        url = "https://baseballsavant.mlb.com/savant-player/{}?stats=career-r-hitting-mlb".format(b)
        res =  requests.get(url)
        soup = BeautifulSoup(res.content, 'lxml')
        for col_name, attribute in zip(attr_names, soup.find_all('div', {'class':'box-text'})):
            attrs[col_name].append(str(attribute.text))
        if idx % 50 == 0:
            with open('../data/baseball_player7.json', 'w') as f:
                json.dump(attrs, f)
        time.sleep(2)
    
    with open('../data/baseball_player7.json', 'w') as f:
        json.dump(attrs, f)
    
    return attrs

#### Webscrape results

The below line of code has been commented out so that I do not run unintentionally and overwrite my already scaped json data. These are the results of the web scrape. I saved the results previously so will load the data in again for use.

In [78]:
# player_info_df = get_batters_info(batter_id)

#### Save Final Scape (json)

The below lines are also commented out so that I do not save over my scrape with any of the changes I will make to the dataframe.

In [ ]:
# player_info_df = pd.DataFrame(player_info_df)
# player_info_df.to_csv('../data/final_scrape.csv')

#### Load scraped data

In [54]:
player_info_df = pd.read_csv('../data/final_scrape.csv')
player_info_df.drop('Unnamed: 0', axis=1, inplace=True)

In [55]:
player_info_df.head()

batter_id height  weight  age position bats throws
0     592450  6' 7"     282   26       RF    R      R
1     519317  6' 6"     245   28       LF    R      R
2     471865  6' 1"     220   32       RF    L      L
3     443558  6' 2"     230   38       DH    R      R
4     121347  6' 3"     230   43       SS    R      R

In [57]:
final_player_info_df = player_info_df[['batter_id', 'height', 'weight', 'age']]

In [58]:
f'Total number of unique players: {len(final_player_info_df.drop_duplicates())}'

'Total number of unique players: 751'

#### Convert all players heights to inches

In [59]:
def height(height_str):
    feet, inches = height_str.split("' ")
    feet = int(feet)
    inches = int(inches[:-1])
    return 12 * feet + inches

In [ ]:
final_player_info_df['height'] = final_player_info_df['height'].map(height)

### Save Data

In [63]:
final_player_info_df.to_csv('../data/final_clean_player_info.csv')

##### Now lets explore the data: 02-EDA